sklearn/utils/_array_api.py

```diff
@@ -0,0 +1,124 @@
+"""Tools to support array_api."""
+import numpy
+from .._config import get_config
+import scipy.special as special
+
+
+class _ArrayAPIWrapper:
+    def __init__(self, array_namespace):
+        self._namespace = array_namespace
+
+    def __getattr__(self, name):
+        return getattr(self._namespace, name)
+
+    def astype(self, x, dtype, *, copy=True, casting="unsafe"):
+        # support casting for NumPy
```

Comment:
```suggestion
        # Extend Array API to support `casting` for NumPy containers
```

Reply:
Is there any issue to track the support of custom casting in the spec?

Reply:
I can not find a discussion on `astype` & `casting`. Maybe @rgommers has a link?

Reply:
I suspect it's because other libraries do not really implement casting for `astype`. For example, [cupy.astype](https://docs.cupy.dev/en/stable/reference/generated/cupy.ndarray.html#cupy.ndarray.astype) does not support casting.

Reply:
For this specific PR, we are using `casting` in `check_array` (which was added in https://github.com/scikit-learn/scikit-learn/pull/14872):

https://github.com/scikit-learn/scikit-learn/blob/cd5385e7112c453afff205fa0ce6a67356cbcf32/sklearn/utils/validation.py#L840

I think we do not need to set `casting` here since the default is unsafe. For reference, the casting behavior of nans and infs [are not specified in the ArrayAPI spec](https://data-apis.org/array-api/latest/API_specification/generated/signatures.data_type_functions.astype.html)

For example:

```python
import numpy as np

print(np.asarray([1, 2, np.nan], dtype=np.float32)
      .astype(np.int32))
# On a x86 machine:
# [          1           2 -2147483648]
# But on a M1 mac:
# [1, 2, 0]

# Cupy cast to zeros.
import cupy
print(cupy.asarray([1, 2, cupy.nan], dtype=cupy.float32).astype(cupy.int32))
# [1, 2, 0]
```

Reply:
> For reference, the casting behavior of nans and infs [are not specified in the ArrayAPI spec](https://data-apis.org/array-api/latest/API_specification/generated/signatures.data_type_functions.astype.html)

That seems like something to specify - even if just to say it's undefined behavior. Which it is I think, as evidenced by the inconsistent NumPy results here.


> I suspect it's because other libraries do not really implement casting for `astype`.

That is typically the reason. The [PR that added `astype`](https://github.com/data-apis/array-api/pull/290) lists all supported keywords across libraries, and only NumPy and Dask have `casting`.

A question is if the concept of casting modes is useful enough to include. I'm not sure to be honest (but I didn't think about it very hard yet). The default in `numpy.ndarray.astype` is unsafe anyway, which is the only reasonable choice probably - because code like `astype(x_f64, float32)` shouldn't raise as it is very explicit.

Reply:
Out of curiosity I did a a quick survey of our current use of the `casting` kwarg.
```
git grep "casting="
sklearn/preprocessing/_polynomial.py:                        casting="no",
sklearn/tree/_tree.pyx:        return n_classes.astype(expected_dtype, casting="same_kind")
sklearn/tree/_tree.pyx:        return value_ndarray.astype(expected_dtype, casting='equiv')
sklearn/tree/_tree.pyx:    return node_ndarray.astype(expected_dtype, casting="same_kind")
sklearn/tree/tests/test_tree.py:    return node_ndarray.astype(new_dtype, casting="same_kind")
sklearn/tree/tests/test_tree.py:    return node_ndarray.astype(new_dtype, casting="same_kind")
sklearn/tree/tests/test_tree.py:    new_n_classes = n_classes.astype(new_dtype, casting="same_kind")
sklearn/utils/validation.py:                    # inf (numpy#14412). We cannot use casting='safe' because
sklearn/utils/validation.py:                    array = array.astype(dtype, casting="unsafe", copy=False)
```

We can ignore the tree files because they are written in Cython and will never benefit from Array API compat.

So we just have `sklearn/preprocessing/_polynomial.py` with `casting="no"` and `sklearn/utils/validation.py` with `casting="unsafe"`.

So it's probably indeed not worth exposing the `casting` argument in our `xp` wrapper and always use `unsafe`.